hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/AbstractBackendEntityEnricher.java

```diff
@@ -0,0 +1,242 @@
+package org.hypertrace.traceenricher.enrichment.enrichers;
+
+import com.typesafe.config.Config;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import javax.annotation.Nullable;
+import org.apache.commons.lang3.tuple.Pair;
+import org.hypertrace.core.datamodel.Event;
+import org.hypertrace.core.datamodel.StructuredTrace;
+import org.hypertrace.core.datamodel.shared.SpanAttributeUtils;
+import org.hypertrace.core.datamodel.shared.StructuredTraceGraph;
+import org.hypertrace.core.datamodel.shared.trace.AttributeValueCreator;
+import org.hypertrace.entity.constants.v1.BackendAttribute;
+import org.hypertrace.entity.data.service.client.EdsClient;
+import org.hypertrace.entity.data.service.v1.AttributeValue;
+import org.hypertrace.entity.data.service.v1.Entity;
+import org.hypertrace.entity.service.constants.EntityConstants;
+import org.hypertrace.semantic.convention.utils.span.SpanSemanticConventionUtils;
+import org.hypertrace.traceenricher.enrichedspan.constants.utils.EnrichedSpanUtils;
+import org.hypertrace.traceenricher.enrichment.AbstractTraceEnricher;
+import org.hypertrace.traceenricher.enrichment.clients.ClientRegistry;
+import org.hypertrace.traceenricher.enrichment.enrichers.cache.EntityCache;
+import org.hypertrace.traceenricher.enrichment.enrichers.resolver.backend.AbstractBackendResolver;
+import org.hypertrace.traceenricher.enrichment.enrichers.resolver.backend.BackendInfo;
+import org.hypertrace.traceenricher.util.EntityAvroConverter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Enricher which gets the backend entities from trace event and store backend entity into
+ * EntityDataService.
+ */
+public abstract class AbstractBackendEntityEnricher extends AbstractTraceEnricher {
```

Comment:
No logic change. Just abstracted `BackendEntityEnricher` into `AbstractBackendEntityEnricher` with 2 additional abstract methods

Reply:
I guess the reason for abstracting this class out is to allow for overriding the set of backend resolvers used? If yes, then wouldn't it better to keep this logic in BackendEntityEnricher and inject the set of resolvers?

Reply:
Yes. That was the motivation. And, I really wanted to keep the logic in `BackendEntityEnricher` and inject the set of resolvers?

But, the question that comes up is how to inject the set of resolvers in `BackendEntityEnricher`?

Reply:
One option is to bind the list of resolvers or we could pass this as config similar to how enrichers are passed as config.

Reply:
That won't be easy. But, I can take a stab at it in a next iteration. That would mean we can't pass `FqnResolver` in the resolver constructor, because, if we have set of resolvers in config, we need to instantiate them via Reflection API, just like enrichers

But, yes, we can pass it via a `setup` function, just like we have `init` function in enrichers

Reply:
> One option is to bind the list of resolvers

Are you hinting towards introducing Dependency Injection in Trace Enricher?

Reply:
Ideally, the set of resolvers should be a sub config of the BackendEntityEnricher config. 

> But, I can take a stab at it in a next iteration. 

Makes sense. We can do it iteratively. But in this pr, are we doing the part around passing resolvers via `setup` function (wherein the Resolvers would be manually instantiated)?

Reply:
> But in this pr, are we doing the part around passing resolvers via setup function (wherein the Resolvers would be manually instantiated)?

I would rather do it as a part of next iteration, and investigate if there are other better ways to instantiate the resolvers. If not, `setup` would be the only way

Reply:
yeah sure

> Are you hinting towards introducing Dependency Injection in Trace Enricher? 
yeah, but the usecase for DI is very less in TraceEnricher

Reply:
> yeah, but the usecase for DI is very less in TraceEnricher

I do see a big use case for DI in Trace Enricher. That would allow us to inject arguments directly via constructors, instead of having `setup`, and `init` functions. Plus, we can inject specific arguments needed for specific enrichers, rather than all enrichers receiving the same arguments via `init`

Though, that's a bigger task, but DI would simplify many use cases